components/async-operation-status-mgt/org.wso2.carbon.identity.framework.async.operation.status.mgt/src/main/java/org/wso2/carbon/identity/framework/async/operation/status/mgt/internal/filter/FilterTreeBuilder.java

```diff
@@ -0,0 +1,285 @@
+                    }
+                    tokenList.add(input.sval);
+                }
+            } else if (input.ttype == '\"' || input.ttype == '\'') {
+                concatenatedString.append(" ").append(input.sval);
+            }
+        }
+        // Add to the list, if the filter is a simple filter.
+        if (!(concatenatedString.toString().equals(""))) {
+            tokenList.add(concatenatedString.toString());
+        }
+    }
+
+    /**
+     * Builds the binary tree from the filterString.
+     *
+     * @return the filter tree as root.
+     */
+    public Node buildTree() throws AsyncOperationStatusMgtClientException {
+
+        expression();
+        return root;
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void expression() throws AsyncOperationStatusMgtClientException {
+
+        term();
+        while (symbol.equals(OR)) {
+            OperationNode or = new OperationNode(OR);
+            or.setLeftNode(root);
+            term();
+            or.setRightNode(root);
+            root = or;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void term() throws AsyncOperationStatusMgtClientException {
+
+        factor();
+        while (symbol.equalsIgnoreCase(AND)) {
+            OperationNode and = new OperationNode(AND);
+            and.setLeftNode(root);
+            factor();
+            and.setRightNode(root);
+            root = and;
+        }
+    }
+
+    /**
+     * We build the parser using the recursive descent parser technique.
+     */
+    private void factor() throws AsyncOperationStatusMgtClientException {
+
+        symbol = nextSymbol();
+        if (symbol.equals(NOT)) {
+            OperationNode not = new OperationNode(NOT);
+            factor();
+            not.setRightNode(root);
+            root = not;
+        } else if (symbol.equals("(")) {
+            expression();
+            symbol = nextSymbol(); // We don't care about ')'.
+        } else {
+            if (!(symbol.equals(")"))) {
+                ExpressionNode expressionNode = new ExpressionNode();
+                validateAndBuildFilterExpression(symbol, expressionNode);
+                root = expressionNode;
+                symbol = nextSymbol();
+            } else {
+                throw handleClientException(ERROR_CODE_INVALID_REQUEST_BODY, symbol);
+            }
+        }
+    }
+
+    /**
+     * Validate the simple filter and build a ExpressionNode
+     *
+     * @param filterString   the filter string.
+     * @param expressionNode the expression node.
+     */
+    private void validateAndBuildFilterExpression(String filterString, ExpressionNode expressionNode)
+            throws AsyncOperationStatusMgtClientException {
+
+        if (StringUtils.isNotBlank(filterString) && !filterString.equals("-1")) {
```

Comment:
```suggestion
        if (StringUtils.isBlank(filterString) || filterString.equals("-1")) {
            return;
        }
```

Use this to avoid nesting